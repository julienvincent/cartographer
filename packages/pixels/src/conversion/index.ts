export * from './pixel-data';
export * from './mc-blocks';
export * from './colors';
