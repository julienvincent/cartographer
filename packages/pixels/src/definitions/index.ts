export * from './palette';
export * from './block';
export * from './pixel';
