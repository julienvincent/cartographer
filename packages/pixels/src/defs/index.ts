export type RGBColor = [number, number, number];

export type Pixel = {
  r: number;
  g: number;
  b: number;
};

export type PixelGrid = Pixel[][];

export type BlockAttributes = Record<string, string> & {
  requires_support?: boolean;
  flammable?: boolean;
  liquid?: boolean;
};

export type MCBlockDefinition = {
  /**
   * The minecraft block ID. This will be something like `minecraft:air`
   */
  id: string;

  /**
   * Stores useful attributes about the block that can be used during the schema block space generation
   */
  attributes?: BlockAttributes;

  /**
   * Stores properties about how the block should be placed. This could be things like `{axis: 'y'}` on a
   * directional block such as `minecraft:birch_log` to indicate the direction the block is facing.
   */
  properties?: Record<string, string>;
};

export type MCBlockWithOffset = MCBlockDefinition & {
  y_offset: number;
};

export type MCBlockGrid = MCBlockWithOffset[][];

export type BlockPaletteItem = MCBlockDefinition & {
  colors: RGBColor[];
};
export type BlockPalette = BlockPaletteItem[];

export type ColorPaletteItem = {
  id: string;
  colors: RGBColor[];
  blocks: MCBlockDefinition[];
};
export type ColorPalette = ColorPaletteItem[];

export enum BlockColorSpectrum {
  Full = 'full',
  Flat = 'flat'
}

export type BlockPaletteItem = MCBlock & {
  colors: PaletteColorTuple;
};
export type BlockPalette = BlockPaletteItem[];
