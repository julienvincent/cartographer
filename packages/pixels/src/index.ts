export * as transformers from './transformers';
export * as conversion from './conversion';
export * as data from './data';
export * as defs from './defs';
