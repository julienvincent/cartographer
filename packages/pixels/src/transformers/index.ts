export * from './color-correction';
export * from './dithering';
