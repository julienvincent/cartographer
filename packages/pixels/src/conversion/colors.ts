import * as nearest from 'nearest-color';
import * as defs from '../definitions';

export const flattenColors = (palette: defs.BlockPalette, spectrum: defs.BlockColorSpectrum) => {
  return palette
    .map((block) => {
      switch (spectrum) {
        case defs.BlockColorSpectrum.Full: {
          return block.colors;
        }
        case defs.BlockColorSpectrum.Flat: {
          return [block.colors[0]];
        }
      }
    })
    .flat()
    .reduce((acc: Record<string, defs.Pixel>, pixel, i) => {
      acc[i] = pixel;
      return acc;
    }, {});
};

export type PixelsToMcColorParams = {
  palette: defs.BlockPalette;

  /**
   * The MC block color spectrum to use when finding matching colors in the palette. This setting will impact how blocks
   * need to be placed in 3D space as the height of blocks relative to each other controls the shade of the color in a
   * MC map.
   *
   * - A value of 'full' will make use of all the colors in the palette resulting in variation in block placement height. This
   *   will result in 'staircasing' in the resulting generated block space - potentially making it harder to build.
   * - A value of 'flat' will use only the first color option in the palette resulting in a perfectly flat map. This will be an
   *   easier build but will have 3x less hue options in the resulting map-art.
   */
  color_spectrum: defs.BlockColorSpectrum;
};

export const convertPixelGridColorsForMC = (pixel_grid: defs.PixelGrid, params: PixelsToMcColorParams) => {
  const findClosestMatchingColor = nearest.from(flattenColors(params.palette, params.color_spectrum));
  return pixel_grid.map((pixels) => {
    return pixels.map((pixel) => {
      const { rgb } = findClosestMatchingColor(pixel);
      return rgb;
    });
  });
};
