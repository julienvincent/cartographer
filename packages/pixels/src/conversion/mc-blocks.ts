import * as defs from '../defs';

const findColorIndex = (colors: defs.RGBColor[], pixel: defs.Pixel) => {
  return colors.findIndex((color) => {
    const [r, g, b] = color;
    return r === pixel.r && g === pixel.g && b === pixel.b;
  });
};

/**
 * Map each pixel in a PixelGrid to it's associated minecraft block id and variant.
 */
export const convertPixelGridToMCBlocks = (
  pixel_grid: defs.PixelGrid,
  palette: defs.BlockPalette
): defs.MCBlockGrid => {
  return pixel_grid.map((pixels) => {
    return pixels.map((pixel) => {
      const block = palette.find((block) => {
        return findColorIndex(block.colors, pixel) !== -1;
      });

      if (!block) {
        throw new Error('could not find matching block');
      }

      return {
        ...block,
        y_offset: findColorIndex(block.colors, pixel)
      };
    });
  });
};
