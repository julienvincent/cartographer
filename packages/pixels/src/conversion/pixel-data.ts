import * as defs from '../defs';
import * as _ from 'lodash';

function average<T>(datums: T[], getter: (datum: T) => number) {
  const sum = datums.reduce((acc, datum) => {
    return acc + getter(datum);
  }, 0);
  return Math.ceil(sum / datums.length);
}

/**
 * Convert a given browser-native ImageData object into a grid of `Pixel`s
 *
 * This essentially converts a flat array of numbers sequentially ordered as r,g,b,a into a
 * multidimensional array of Pixel maps.
 *
 * [r,g,b,a,  r,g,b,a,  r,g,b,a,  r,g,b,a]
 * ->>
 * [[Pixel, Pixel], [Pixel, Pixel]]
 *
 * This makes working with the data in later steps much easier
 */
export const convertImageDataToPixelGrid = (image_data: ImageData): defs.PixelGrid => {
  const pixels = _.chunk(image_data.data, 4).map((pixel) => {
    const [r, g, b] = pixel;
    return {
      r,
      g,
      b
    };
  });

  return _.chunk(pixels, image_data.width);
};

export const convertPixelGridToImageData = (pixel_grid: defs.PixelGrid) => {
  const width = pixel_grid[0].length;
  const height = pixel_grid.length;

  return pixel_grid.reduce((image_data, pixels, h) => {
    return pixels.reduce((image_data, pixel, w) => {
      const i = h * width * 4 + w * 4;
      const { r, g, b } = pixel;
      image_data.data[i] = r;
      image_data.data[i + 1] = g;
      image_data.data[i + 2] = b;
      image_data.data[i + 3] = 255;
      return image_data;
    }, image_data);
  }, new ImageData(width, height));
};

/**
 * Attempt to scale a given pixel grid down to a given target width/height. This works by creating
 * quadrants in the source pixel grid and averaging the pixels in each quadrant together
 */
export const scaleDownPixelGrid = (pixel_grid: defs.PixelGrid, width: number, height: number) => {
  const source_width = pixel_grid[0].length;
  const source_height = pixel_grid.length;

  const x_scale = source_width / width;
  const y_scale = source_height / height;

  return _.range(height).map((target_y) => {
    return _.range(width).map((target_x) => {
      const pixels: defs.Pixel[] = [];

      const source_y_origin = target_y === 0 ? y_scale : target_y * y_scale;
      const source_x_origin = target_x === 0 ? x_scale : target_x * x_scale;

      const range_min_y = Math.round(source_y_origin - y_scale / 2);
      const range_max_y = Math.round(source_y_origin + y_scale / 2);

      const range_min_x = Math.round(source_x_origin - x_scale / 2);
      const range_max_x = Math.round(source_x_origin + x_scale / 2);

      for (let source_y = range_min_y; source_y < range_max_y; source_y++) {
        for (let source_x = range_min_x; source_x < range_max_x; source_x++) {
          const y_diff_factor = 1 - Math.max(0, Math.abs(source_y - source_y_origin) - range_max_y);
          const x_diff_factor = 1 - Math.max(0, Math.abs(source_x - source_x_origin) - range_max_x);

          const { r, g, b } = pixel_grid[source_y][source_x];
          pixels.push({
            r: r * y_diff_factor * x_diff_factor,
            g: g * y_diff_factor * x_diff_factor,
            b: b * y_diff_factor * x_diff_factor
          });
        }
      }

      return {
        r: average(pixels, (pixel) => pixel.r),
        g: average(pixels, (pixel) => pixel.g),
        b: average(pixels, (pixel) => pixel.b)
      };
    });
  });
};

export const scaleUpPixelGrid = (pixel_grid: defs.PixelGrid, width: number, height: number) => {
  const source_width = pixel_grid[0].length;
  const source_height = pixel_grid.length;
  const width_pixel_scale = width / source_width;
  const height_pixel_scale = height / source_height;

  return _.range(height).map((target_height) => {
    return _.range(width).map((target_width) => {
      const pixel_w = Math.floor(target_width / width_pixel_scale);
      const pixel_h = Math.floor(target_height / height_pixel_scale);
      return pixel_grid[pixel_h][pixel_w];
    });
  });
};
