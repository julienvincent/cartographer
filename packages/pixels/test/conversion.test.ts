import { it, expect, describe } from 'vitest';

import * as conversion from '../src/conversion';

describe('pixel conversions', () => {
  it('should correctly convert ImageData to a pixel grid', () => {});

  it('should correctly scale a PixelGrid down', () => {
    const arr = [1, 2];

    expect(arr).toMatchSnapshot();
  });
});
