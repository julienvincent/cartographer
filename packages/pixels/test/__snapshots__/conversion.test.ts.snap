// Vitest Snapshot v1

exports[`pixel conversions > should correctly scale a PixelGrid down 1`] = `
[
  1,
  2,
]
`;
